### ISSUE TITLE
`TextImageDataset` bottlenecks

### ISSUE THREAD

**Issue by afiaka87** [lucidrains/DALLE-pytorch#216]
## The init method

The init method is quite fast - but when dealing with more than 100k image-text pairs, it can take quite awhile.

These few lines result in full scans over the length of both the `image_files` list and the `text_files` list a couple of times:

```python
        path = Path(folder)

        text_files = [*path.glob('**/*.txt')]
        image_files = [
            *path.glob('**/*.png'), *path.glob('**/*.jpg'),
            *path.glob('**/*.jpeg'), *path.glob('**/*.bmp')
        ]

        text_files = {text_file.stem: text_file for text_file in text_files} # full scan
        image_files = {image_file.stem: image_file for image_file in image_files} # full scan
        keys = (image_files.keys() & text_files.keys()) # Probably okay
        self.keys = list(keys) 
        self.text_files = {k: v for k, v in text_files.items() if k in keys} # Full scan for "matching"
        self.image_files = {k: v for k, v in image_files.items() if k in keys}  # Full scan for "matching
```

Is there any way to speed this up? Or at least give the user an option to cache the lists into a file for resuming?

## Smaller optimizations

Other parts of the data loader may bottleneck on slower configurations (although it hasn't been an issue really):

If the user isn't shuffling, **and** doesn't mind a Dataloader with less "variety" in it, you can actually cache the "sequential_sample" method like so -

```python
# loader.py
from functools import lru_cache
# ...
    @lru_cache(maxsize=None)
    def sequential_sample(self, ind):
        if ind >= self.__len__() - 1:
            return self.__getitem__(0)
        return self.__getitem__(ind + 1)
```

This way, you only load the tensors for the text and image once on your first epoch. 

This will however eliminate any "side effects" used in the method. As such, the value randomly chosen for your "resize_ratio" will remain the same for that tensor during each epoch. 

If you were _really_ crazy, and wanted to lose even more variety in your dataset, you could do something similar to the `random_sample()` method. This has the added benefit (?) of caching the corrupted indices of your dataloader and returning another (cached) image-text-pair in your dataset instead. You can raise/lower your maxsize to increase/decrease the amount of cache hits/misses and therefore the frequency of side effects - which could be tinkered with to find a good value which helps your loss and the bottleneck.

```python
    @lru_cache(maxsize=4096)
    def random_sample(self):
        return self.__getitem__(randint(0, self.__len__() - 1))
```

I _hope_ this is already happening in the background via either Python or `torch.Dataset`. One thing that couldn't possibly hurt however, would be to add an `lru_cache(maxsize=1)` to the `__len__()` method. 

```python
    @lru_cache(maxsize=1)
    def __len__(self):
        return len(self.keys)
```

**Comment by robvanvolt** [comment]
Maybe we should convert our datasets to tar.gz and work with WebDatasetloaders? This might be a faster approach than optimizing reading text and image folders, doesn't it? But cool anyway to get some extra speed of the current approach! (Y)

**Comment by afiaka87** [comment]
> Maybe we should convert our datasets to tar.gz and work with WebDatasetloaders? This might be a faster approach than optimizing reading text and image folders, doesn't it? But cool anyway to get some extra speed of the current approach! (Y)

I agree. WebDataset is perfect for our usecase.

**Comment by robvanvolt** [comment]
Added early beta support for WebDatasets:
https://github.com/lucidrains/DALLE-pytorch/pull/280

**Comment by robvanvolt** [comment]
> Added early beta support for WebDatasets:
> #280

Added full support for WebDatasets: https://github.com/lucidrains/DALLE-pytorch/pull/280
Anyone wants to try the new feature out or review the changes?
